### ISSUE TITLE
About weight pruning and entropy coding

### ISSUE THREAD

**Issue by JoochanLee** [haochen-rye/NeRV#16]
https://github.com/haochen-rye/NeRV/blob/adf61b81fc192c64d2de7b93745b28ff1cf33a39/train_nerv.py#L442

When compressing the network's weights by Huffman codes, I confirmed that zero values are excepted.
In this case, we can not know the position of pruned weights when reconstructing the model weights.
I think additional information(such as indices of pruned weight) is required.

Can you explain the details?

Thank you for sharing the brilliant work!

**Comment by haochen-rye** [comment]
Before entropy coding, we have a model pruning step and only non-zero weights will be remained after the pruning step. So we remove them in the entropy coding step.

**Comment by JoochanLee** [comment]
> 
When decoding the compressed model according to the code stream without 0's, I think we can't reconstruct the model because we don't know the position of deactivated weights.

If applying a constructed pruning (layer or channel pruning), storage overhead for saving index is not significant.
However, the index overhead of unconstructed pruning is not negligible. (Please refer to https://arxiv.org/pdf/1907.02124.pdf)

Do you suppose a client already knows which weights are deactivated?

**Comment by haochen-rye** [comment]
Yeah, I think you are right. How to store the sparse model weights for non-structure pruning should be discussed in more details.  Actually I have been thinking of structure pruning and not as a post-process bur as a structure regularization on model weights during training, in which case we do not need extra pruning steps and can prune channels and don't need to stored the sparse positions/indexes). 
But as you you said, I should clarify this in the paper and code!
